### PR TITLE
Add a new FAQ

### PR DESCRIPTION
**Pull request by sanjoy**
"What is the difference between traits and operation interfaces"

### CODE REVIEW

File: website/content/getting_started/Faq.md [hunk 1]
```diff
@@ -102,3 +102,21 @@ Attribute getConstantAttr(Operation *constantOp) {
   return constant;
 }
 ```
+
+## What is the difference between traits and operation interfaces?
```
Comment (River707) [inline website/content/getting_started/Faq.md:106]: I would drop the `operation` here (or reword), the same applies for attributes and type traits/interfaces.

File: website/content/getting_started/Faq.md [hunk 2]
```diff
@@ -102,3 +102,21 @@ Attribute getConstantAttr(Operation *constantOp) {
...
+
+Both [traits](https://mlir.llvm.org/docs/Traits/) and [operation
```
Comment (River707) [inline website/content/getting_started/Faq.md:108]: Same here. Might want to slight reword references to Op throughout.

**Review by River707 — CHANGES_REQUESTED** [review]

File: website/content/getting_started/Faq.md [hunk 3]
```diff
@@ -102,3 +102,21 @@ Attribute getConstantAttr(Operation *constantOp) {
...
+interfaces](https://mlir.llvm.org/docs/Interfaces/#attributeoperationtype-interfaces)
+can be used to inject common behavior into various operations without
+duplication.  However, conceptually these are quite different.
+
+Traits inject static behavior into `Op`s whereas operation interfaces
+dynamically dispatch behavior based on the operation type.  Operation interfaces
+need a vtable or equivalent while traits don't.  In other words, traits are like
+C++ base classes without virtual methods while op interfaces are like C++ base
+classes with virtual methods.
```
Comment (joker-eph) [inline website/content/getting_started/Faq.md:117]: I'm not sure I understand the analogy with a C++ base class without virtual methods.
For example such C++ base class still allows for polymorphism, that is:

```
void f(BaseClass &b) {
  b.foo(); // calling a non virtual method of the base class independently of the actual operation.
}
```

I don't think a Trait allows for this?

File: website/content/getting_started/Faq.md [hunk 4]
```diff
@@ -102,3 +102,21 @@ Attribute getConstantAttr(Operation *constantOp) {
...
+## What is the difference between traits and operation interfaces?
+
+Both [traits](https://mlir.llvm.org/docs/Traits/) and [operation
+interfaces](https://mlir.llvm.org/docs/Interfaces/#attributeoperationtype-interfaces)
+can be used to inject common behavior into various operations without
+duplication.  However, conceptually these are quite different.
+
+Traits inject static behavior into `Op`s whereas operation interfaces
+dynamically dispatch behavior based on the operation type.  Operation interfaces
+need a vtable or equivalent while traits don't.  In other words, traits are like
```
Comment (bondhugula) [inline website/content/getting_started/Faq.md:115]: I'm not sure it's useful to reference `vtable` -- that just comes with standard OO polymorphism and is a detail?

**Comment by bondhugula** [comment]
This FAQ is quite useful to have - thanks!

File: website/content/getting_started/Faq.md (hunk 4 above)
Comment (sanjoy) [reply website/content/getting_started/Faq.md:115]: Removed analogy with C++.

File: website/content/getting_started/Faq.md (hunk 3 above)
Comment (sanjoy) [reply website/content/getting_started/Faq.md:117]: Removed analogy with C++.

**Comment by sanjoy** [comment]
Addressed comments.

**Review by joker-eph — APPROVED** [review]
Thanks! That's a great blob, I love the examples :)

**Pull request merged**